samples/enum.golo

```diff
@@ -0,0 +1,109 @@
+# Copyright 2012-2015 Institut National des Sciences Appliquées de Lyon (INSA-Lyon)
+# 
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# 
+#     http://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module samples.Enums
+
+enum Option = {
+  Some = { value }
+  None
+}
+
+augment Option {
+  function bind = |this, func| -> match {
+    when this is Option.None() then this
+    otherwise func(this: value())
+  }
+
+  function fmap = |this, func| -> match {
```

Comment:
To match `java.util.Optional`, `bind` should be `flatMap` and `fmap` should be `map`. I used here the classical monad vocabulary, but I can switch to Java if you prefer